### PR TITLE
Queue discrete events in microtask

### PR DESCRIPTION
**Pull request by rickhanlonii**
## Overview

This PR is part of this stack:

- [Improve tests that use discrete events](https://github.com/facebook/react/pull/20667)
- [Add queue microtask to host configs](https://github.com/facebook/react/pull/20668)
- [Queue discrete events in microtask (this PR)](https://github.com/facebook/react/pull/20669)

As a concurrent mode experiment, we want try reverting back to flushing discrete events synchronously so that the upgrade path is easier. This will also help users catch bugs and make the behavior more consistent; since we're going to schedule the sync flush in a microtask instead of within the same stack frame. This ensures that the update is "async", but is still flushed synchronously at the end of the current task, in order, before other work.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 1]
```diff
@@ -726,6 +728,13 @@ function ensureRootIsScheduled(root: FiberRoot, currentTime: number) {
       ImmediateSchedulerPriority,
       performSyncWorkOnRoot.bind(null, root),
     );
+  } else if (
+    enableDiscreteEventMicroTasks &&
+    queueMicrotask != null &&
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:733]: Nit: I think this will get inlined properly but just to be sure, instead of making `queueMicrotask` undefined for some host configs, you can throw an error (`throw new Error('Not implemented')`) in the ones where `enableDiscreteEventMicroTasks` is meant to disabled.

I think that's sufficient because you're turning this on everywhere except the www feature flags.

If you do leave it this way, please confirm the type check isn't in the production bundle.

**Review by acdlite — APPROVED** [review]

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 2]
```diff
@@ -726,6 +728,13 @@ function ensureRootIsScheduled(root: FiberRoot, currentTime: number) {
       ImmediateSchedulerPriority,
       performSyncWorkOnRoot.bind(null, root),
     );
+  } else if (
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:731]: GitHub won't let me comment on the lines above, but need to change 

```js
if (existingCallbackNode !== null) {
```

to check the priority instead, like we discussed out-of-band:

```js
if (existingCallbackPriority !== NoPriority) {
```

That will prevent it from scheduling a separate microtask per individual update

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:731]: Ah yeah, I still had that in my notes. Will do.

File: packages/react-dom/src/client/ReactDOMHostConfig.js [hunk 3]
```diff
@@ -384,6 +384,21 @@ export const scheduleTimeout: any =
...
 export const noTimeout = -1;
+export const queueMicrotask: (Function => void) | undefined =
+  typeof queueMicrotask === 'function'
+    ? queueMicrotask
+    : typeof Promise !== 'undefined'
+    ? callback =>
+        Promise.resolve(null)
+          .then(callback)
+          .catch(handleErrorInNextTick)
+    : scheduleTimeout;
```
Comment (acdlite) [inline packages/react-dom/src/client/ReactDOMHostConfig.js:395]: Let's leave a TODO here to consider the best fallback. I think we need to warn consumers to use a Promise polyfill.

And then as a maybe as a last resort it should schedule a Scheduler task.

Comment (acdlite) [reply packages/react-dom/src/client/ReactDOMHostConfig.js:395]: Added item to the umbrella

Comment (rickhanlonii) [reply packages/react-dom/src/client/ReactDOMHostConfig.js]: Added TODO

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js (hunk 1 above)
Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js]: Done

**Comment by acdlite** [comment]
Updates look great! Thanks!

File: packages/react-reconciler/src/forks/ReactFiberHostConfig.custom.js [hunk 4]
```diff
@@ -54,6 +54,7 @@ export const shouldSetTextContent = $$$hostConfig.shouldSetTextContent;
 export const createTextInstance = $$$hostConfig.createTextInstance;
 export const scheduleTimeout = $$$hostConfig.scheduleTimeout;
 export const cancelTimeout = $$$hostConfig.cancelTimeout;
+export const queueMicrotask = $$$hostConfig.queueMicrotask;
```
Comment (gaearon) [inline packages/react-reconciler/src/forks/ReactFiberHostConfig.custom.js:57]: Can you please briefly document it in the `react-reconciler` README?

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 5]
```diff
@@ -710,33 +712,68 @@ function ensureRootIsScheduled(root: FiberRoot, currentTime: number) {
...
 
-  // Check if there's an existing task. We may be able to reuse it.
-  if (existingCallbackNode !== null) {
+    // Check if there's an existing task. We may be able to reuse it.
     const existingCallbackPriority = root.callbackPriority;
-    if (existingCallbackPriority === newCallbackPriority) {
-      // The priority hasn't changed. We can reuse the existing task. Exit.
+    if (
+      existingCallbackPriority !== NoLanePriority &&
+      existingCallbackPriority !== InputDiscreteLanePriority
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:730]: I just noticed this check, why is this necessary? Checking `existingCallbackNode` is sufficient. If you want the invariant I would make that a DEV-only warning so we don't do an extra check in prod.

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js]: I thought the goal was to not check existingCallbackNode and specify the priorities that shouldn't have a callback scheduled?

File: packages/react-reconciler/src/forks/ReactFiberHostConfig.custom.js (hunk 4 above)
Comment (rickhanlonii) [reply packages/react-reconciler/src/forks/ReactFiberHostConfig.custom.js]: Done 👍

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 6]
```diff
@@ -716,21 +716,27 @@ function ensureRootIsScheduled(root: FiberRoot, currentTime: number) {
...
-  if (existingCallbackNode !== null) {
-    const existingCallbackPriority = root.callbackPriority;
-    if (existingCallbackPriority === newCallbackPriority) {
-      // The priority hasn't changed. We can reuse the existing task. Exit.
-      return;
+  const existingCallbackPriority = root.callbackPriority;
+  if (existingCallbackPriority === newCallbackPriority) {
+    if (__DEV__) {
+      invariant(
+        existingCallbackNode,
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:730]: This seems wrong? `existingCallbackNode` will be null if the existing task is a microtask.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:730]: Side note: We shouldn't have dev-only invariants. Use a warning so the semantics don't diverge. Though because of my previous point, the warning shouldn't exist, either.

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 7]
```diff
@@ -716,21 +716,27 @@ function ensureRootIsScheduled(root: FiberRoot, currentTime: number) {
...
+      );
     }
-    // The priority changed. Cancel the existing callback. We'll schedule a new
-    // one below.
+    // The priority hasn't changed. We can reuse the existing task. Exit.
+    return;
+  }
+
+  if (existingCallbackNode != null) {
+    // Cancel the existing callback. We'll schedule a new one below.
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:739]: I think the warning you wanted goes in the else block of this check.

```js
if (existingCallbackNode != null) {
  // Cancel the existing callback. We'll schedule a new one below.
  cancelCallback(existingCallbackNode);
} else {
  if (__DEV__ && newCallbackPriority !== InputDiscreteLanePriority && newCallbackPriority !==  SyncLanePriority) {
    console.error('oh noes!');
  }
} 
```

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:739]: If it goes in this block, then we may have already returned without something being scheduled.

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js (hunk 6 above)
Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:730]: Yeah, was just fixing this.

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js (hunk 7 above)
Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:739]: Also, there will be no callback on the first schedule.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:739]: One more nit: generally, for our own internal types, we always explicitly compare to `null`. That way 1) a missing field/argument is distinct from one that is intentionally null, and 2) it's a tiny bit faster.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:742]: > If it goes in this block, then we may have already returned without something being scheduled.

Can you rephrase? The check I'm suggesting you move is already behind a __DEV__  check which means it doesn't affect behavior. If it did then we'd definitely want to remove the __DEV__ check :D

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:742]: I thought so, but then there were a bunch of invariant calls that didn't do a null check so I switched it.

Fixed 👍

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:742]: Yeah, if we move the check from the block above this here, and the control flow returns in the above block, then it will never hit this invariant check. Which means we would chose to reuse and exiting task that doesn't exist.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.new.js:742]: Ok you can leave that there then but please make it a warning, not an error.

I was thinking you wanted to warn about a slightly different case, which is if you expect to cancel a non-microtask task but the callback node doesn't exist as expected.

File: scripts/error-codes/codes.json [hunk 8]
```diff
@@ -372,5 +372,6 @@
   "381": "This feature is not supported by ReactSuspenseTestUtils.",
   "382": "This query has received more parameters than the last time the same query was used. Always pass the exact number of parameters that the query needs.",
   "383": "This query has received fewer parameters than the last time the same query was used. Always pass the exact number of parameters that the query needs.",
-  "384": "Refreshing the cache is not supported in Server Components."
+  "384": "Refreshing the cache is not supported in Server Components.",
+  "385": "Expected scheduled callback to exist. This error is likely caused by a bug in React. Please file an issue."
```
Comment (acdlite) [inline scripts/error-codes/codes.json:376]: This can go away now
